## Упражнения по библиотеке Numpy

In [16]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
random_array = np.random.randint(0, 10, (3, 7))
new_array = np.where((random_array>3) & (random_array<8), -random_array, random_array)

print(f"Исходный массив:\n {random_array}")
print(f"Измененный массив:\n {new_array}")

Исходный массив:
 [[1 8 7 2 3 5 8]
 [2 8 2 4 2 9 6]
 [1 0 6 3 5 2 2]]
Измененный массив:
 [[ 1  8 -7  2  3 -5  8]
 [ 2  8  2 -4  2  9 -6]
 [ 1  0 -6  3 -5  2  2]]


**2.** Заменить максимальный элемент случайного массива на 0

In [3]:
random_array_2 = np.random.randint(0, 10, (3, 7))
new_array_2 = np.where((random_array_2 == np.max(random_array_2)), 0, random_array_2)

print(f"Исходный массив:\n {random_array_2}")
print(f"Измененный массив:\n {new_array_2}")

Исходный массив:
 [[6 5 0 6 6 4 2]
 [2 0 4 2 2 6 5]
 [4 8 4 5 8 8 8]]
Измененный массив:
 [[6 5 0 6 6 4 2]
 [2 0 4 2 2 6 5]
 [4 0 4 5 0 0 0]]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [85]:
matrix = np.random.randint(1,9,(3,2))

result_gen = np.array(np.meshgrid(*matrix)).T.reshape(-1, matrix.shape[0])

print(f"Исходный массив:\n {matrix}")
print(f"Прямое произведение:\n {result_gen}")

Исходный массив:
 [[5 1]
 [2 8]
 [5 1]]
Прямое произведение:
 [[5 2 5]
 [5 8 5]
 [1 2 5]
 [1 8 5]
 [5 2 1]
 [5 8 1]
 [1 2 1]
 [1 8 1]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [51]:
A = np.random.randint(0,9,(8,3))
B = np.random.randint(0,9,(2,2))

print(f"Массив А:\n {A}")
print (f"Массив B:\n {B}")

Массив А:
 [[0 1 4]
 [6 3 0]
 [5 8 0]
 [3 2 5]
 [7 3 5]
 [8 3 7]
 [4 7 6]
 [4 5 0]]
Массив B:
 [[0 6]
 [7 4]]


In [52]:
mask = np.array([[np.all(np.isin(B_row, A_row)) for B_row in B]
    for A_row in A
])

A[mask.any(axis=1)]

array([[6, 3, 0],
       [4, 7, 6]])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [5]:
arr = np.array([[3, 8, 9],
                [4, 4, 4],
                [6, 3, 8], 
                [7, 7, 7], 
                [5, 1, 3], 
                [9, 5, 6], 
                [2, 2, 2], 
                [1, 9, 7],
                [9, 9, 9], 
                [2, 5, 6]])

mask = np.all(arr == arr[:, [0]], axis = 1)

print(f"Массив со строками из неравных значений:\n\n {arr[~mask]}")

Массив со строками из неравных значений:

 [[3 8 9]
 [6 3 8]
 [5 1 3]
 [9 5 6]
 [1 9 7]
 [2 5 6]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [34]:
arr = np.array([[2, 3, 4, 6],
                [14, 27, 11, 6],
                [2, 3, 4, 6],
                [12, 33, 24, 60],
                [2, 3, 4, 6],
                [12, 33, 24, 60],
                [1, 13, 24, 10],
                [1, 13, 24, 10]])

print(f"Массив без дубликатов:\n\n {np.unique(arr, axis =0)}")

Массив без дубликатов:

 [[ 1 13 24 10]
 [ 2  3  4  6]
 [12 33 24 60]
 [14 27 11  6]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [3]:
#вариант 1

import random

random_list_of_lists = [[random.randint(0, 10) for _ in range(3)] for _ in range(5)]

x_prod = 1
for i, inner_list in enumerate(random_list_of_lists):
    for j, item in enumerate(inner_list):
        if i == j and item != 0:
            x_prod *= item
print(f"Исходный список: \n {random_list_of_lists}")            
print(f"Произведение ненулевых элементов на главной диагонали: {x_prod}")

Исходный список: 
 [[8, 1, 9], [7, 5, 10], [8, 0, 7], [4, 8, 6], [9, 7, 10]]
Произведение ненулевых элементов на главной диагонали: 280


In [98]:
# вариант 2

X = np.random.randint(0,9,(5,3))

diagonal_elements = np.diag(X)
diag_prod = np.prod(diagonal_elements[diagonal_elements != 0])

print(f"Исходная матрица:\n {X}")
print(f"Произведение ненулевых элементов на главной диагонали = {diag_prod}")

Исходная матрица:
 [[6 8 2]
 [3 3 0]
 [3 1 1]
 [3 2 5]
 [8 1 7]]
Произведение ненулевых элементов на главной диагонали = 18


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [18]:
#вариант 1

from collections import Counter

x = [random.randint(0, 10) for _ in range(5)]
y = [random.randint(0, 10) for _ in range(5)]

print(f"Исходные векторы: \n {x} \n {y}")

print(Counter(x) == Counter(y))    

Исходные векторы: 
 [3, 3, 10, 3, 3] 
 [2, 0, 8, 4, 8]
False


In [20]:
#вариает 2

x_arr = np.random.randint(0, 10, 5)
y_arr = np.random.randint(0, 10, 5)

print(f"Исходные векторы: \n {x_arr} \n {y_arr}")

print(np.array_equal(np.bincount(x_arr), np.bincount(y_arr)))

Исходные векторы: 
 [8 2 4 3 8] 
 [4 6 8 2 8]
False


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [119]:
#вариант 1
x_3 = [random.randint(0, 10) for _ in range(12)]

max_value = max((x_3[i] for i in range(1, len(x_3)) if x_3[i-1] == 0), default=None)

print(f"Исходный вектор: {x_3}")
print(f"Искомое максимальное значение: {max_value}")

Исходный вектор: [7, 10, 0, 0, 10, 1, 0, 5, 8, 0, 7, 6]
Искомое максимальное значение: 10


In [120]:
#вариант 2

x_arr_3 = np.random.randint(0, 10, 12)

x_arr_3_new = np.roll(x_arr_3, 1)
elements_after_zero = x_arr_3[x_arr_3_new == 0]
max_value = elements_after_zero.max() if elements_after_zero.size > 0 else None

print(f"Исходный вектор: {x_arr_3}")
print(f"Искомое максимальное значение: {max_value}")

Исходный вектор: [8 1 4 9 2 7 0 4 7 8 4 6]
Искомое максимальное значение: 4


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [38]:
#вариант 1

x_4 = [random.randint(0, 10) for _ in range(15)]
unique_values = list(set(x_4))

counts = [x_4.count(n) for n in unique_values]

print(f"Исходный вектор: {x_4}")
print(f"Результат Run-length encoding: {(unique_values, counts)}")

Исходный вектор: [3, 4, 8, 5, 3, 0, 9, 5, 9, 8, 6, 9, 3, 2, 8]
Результат Run-length encoding: ([0, 2, 3, 4, 5, 6, 8, 9], [1, 1, 3, 1, 2, 1, 3, 3])


In [39]:
#вариант 2

x_arr_4 = np.random.randint(0, 10, 15)
u_values, elem_counts = np.unique(x_arr_4, return_counts=True)

print(f"Исходный вектор: {x_arr_4}")
print(f"Результат Run-length encoding: {(u_values, elem_counts)}")

Исходный вектор: [9 0 9 2 9 8 0 2 5 8 6 7 5 7 8]
Результат Run-length encoding: (array([0, 2, 5, 6, 7, 8, 9]), array([2, 2, 2, 1, 2, 3, 3], dtype=int64))


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [105]:
import timeit
import math
from scipy.spatial.distance import cdist

In [106]:
#вариант 1

X_5 = [[random.randint(0, 10) for _ in range(2)] for _ in range(3)]
Y_5 = [[random.randint(0, 10) for _ in range(2)] for _ in range(2)]

start_time = timeit.timeit()

с = []

for x_5 in X_5:
    row = [math.sqrt(sum((xi - yi) ** 2 for xi, yi in zip(x_5, y_5))) for y_5 in Y_5]
    distance_matrix.append(row)
    
end_time = timeit.timeit()
execution_time_var_1 = start_time-end_time

    
print(f"Матрица евклидовых расстояний между объектами:\n {distance_matrix}")
print(f"Скорость работы: {execution_time_var_1}")

Матрица евклидовых расстояний между объектами:
 [[8.06225774829855, 5.0], [4.47213595499958, 0.0], [2.23606797749979, 3.0]]
Скорость работы: 0.001519199984613806


In [113]:
#вариант 2

X_arr_5 = np.random.randint(0, 10, (3, 2))
Y_arr_5 = np.random.randint(0, 10, (2,2))

np.sqrt(np.sum((X_arr_5[:, np.newaxis] - Y_arr_5) ** 2, axis=-1))
execution_time_var_2 = timeit.timeit('np.sqrt(np.sum((X_arr_5[:, np.newaxis] - Y_arr_5) ** 2, axis=-1))',globals=globals(), number=10)

print(f"Матрица евклидовых расстояний между объектами:\n {np.sqrt(np.sum((X_arr_5[:, np.newaxis] - Y_arr_5) ** 2, axis=-1))}")
print(f"Скорость работы: {execution_time_var_2}")
print(f"Скорость работы cdist: {timeit.timeit('cdist(X_arr_5, Y_arr_5)',globals=globals(), number=10)}")

Матрица евклидовых расстояний между объектами:
 [[3.         5.38516481]
 [7.61577311 5.09901951]
 [8.06225775 6.08276253]]
Скорость работы: 0.00031689999741502106
Скорость работы cdist: 0.005135700019309297


Вычисление матрицы евклидовых расстояний между объектами с использованием numpy работает быстрее, чем функция scipy.spatial.distance.cdist

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [77]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [40]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [116]:
average_calories = np.mean(calorie_stats)
print(f"Среднее количество калорий в хлопьях у конкурентов {average_calories:.2f}, что выше на {average_calories-60:.2f}, чем CrunchieMunchies ")

Среднее количество калорий в хлопьях у конкурентов 106.88, что выше на 46.88, чем CrunchieMunchies 


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [45]:
calorie_stats_sorted = np.sort(calorie_stats)
print(f"Отсортированные данные:\n {calorie_stats_sorted}")

Отсортированные данные:
 [ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [48]:
median_calories = np.median(calorie_stats)
print(f"Медиана: {median_calories}")

Медиана: 110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [118]:
percentile = [0,3, 4, 5, 7, 8, 9, 10, 25, 50, 75, 90, 95, 99]
nth_percentile_search = np.percentile(calorie_stats, percentile)

nth_percentile = None

for p, value in zip(percentile, nth_percentile_search):
    if value > 60 and nth_percentile is None:
        nth_percentile = p   
    print(f"{p}%-й процентиль: {value}")
    
print(f"\nНаименьший процентиль, превышающий 60 калорий: {nth_percentile}%-й процентиль")

0%-й процентиль: 50.0
3%-й процентиль: 55.599999999999994
4%-й процентиль: 70.0
5%-й процентиль: 70.0
7%-й процентиль: 83.2
8%-й процентиль: 90.0
9%-й процентиль: 90.0
10%-й процентиль: 90.0
25%-й процентиль: 100.0
50%-й процентиль: 110.0
75%-й процентиль: 110.0
90%-й процентиль: 124.00000000000006
95%-й процентиль: 140.0
99%-й процентиль: 152.39999999999995

Наименьший процентиль, превышающий 60 калорий: 4%-й процентиль


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [84]:
more_calories = len(calorie_stats[calorie_stats > 60]) / len(calorie_stats) * 100
print(f"Процент хлопьев, содержащих более 60 калорий: {more_calories:.2f}")

Процент хлопьев, содержащих более 60 калорий: 96.10


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [86]:
calorie_std = np.std(calorie_stats)
print(f"Стандартное отклонение: {calorie_std}")

Стандартное отклонение: 19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Калорийность хлопьев CrunchieMunchies в среднеем ниже на 46.88 калорий, чем у конкурентов. Но сравнение со средним значением нельзя считать корректным, так как медианное значение калорийности хлопьев конкурентов первышает среднее, а значит отличие  CrunchieMunchies еще больше (50 калорий). Также  96% исследованных производителей выпускают хлопья с калорийностью выше, чем CrunchieMunchies, т.е. наш продукт входит в 4% самых низкокалорийных, представленных на рынке.